aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -377,7 +376,89 @@ Tensor logsumexp(const Tensor &self, int64_t dim_, bool keepdim) {
 
 // MULTI DIM REDUCE ###########################################################
 
-template <Tensor (reduce_1)(const Tensor &, int64_t, bool)>
+// NB: this applies two optimizations:
+//   1. Reducing the dimensions in the order of decreasing size, so that the
+//      larger dimensions are dealt earlier and we can work with less elements
+//      overall.
```

Comment:
It's a neat idea but are you sure it won't mess up with locality? Remember that summing over contiguous dimensions is a few times faster. What if you tried to sum a tensor like this one:

```python
torch.randn(10, 200, 100).permute(0, 2, 1)
```

In general, since the `_out` output will be contiguous for sure, it might be beneficial to go for the contig dim first, and only then use this sorting.

Reply:
@apaszke Good point! I will do some benchmarking tomorrow.